File: algorithms/cpp/16.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int threeSumClosest(vector<int>& nums, int target)
    {
        int result = nums[0]+nums[1]+nums[2];
        sort(nums.begin(), nums.end());
        for ( int i = 0; i < (int)nums.size()-2; i++ )
        {
            int j = i+1, k = nums.size()-1;
            while ( j < k )
            {
                int sum = nums[i]+nums[j]+nums[k];
                if ( sum == target )
                    return target;
                result = abs(result-target)>abs(sum-target) ? sum : result;
                if ( sum < target )
                    j++;
                else
                    k--;
            }
        }
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/python3/16.py
from bisect import bisect_left

class Solution:
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        rls = nums[0] + nums[1] + nums[2]
        for i in range(len(nums)-2):
            for j in range(i+1, len(nums)-1):
                t = target - nums[i] - nums[j]
                idx = bisect_left(nums, t, j+1)
                if j < idx < len(nums):
                    s = nums[i] + nums[j] + nums[idx]
                    if abs(rls - target) > abs(s - target):
                        rls = s
                if j < idx-1 < len(nums):
                    s = nums[i] + nums[j] + nums[idx-1]
                    if abs(rls - target) > abs(s - target):
                        rls = s
        return rls
